src/row_codec.rs

```diff
@@ -1,182 +1,230 @@
-    data.advance(key_suffix_len);
-
-    let meta = decode_meta(row_features, data)?;
-    let value = if !(meta.flags & RowFlags::Tombstone).is_empty() {
-        ValueDeletable::Tombstone
-    } else {
-        let value_len = data.get_u32() as usize;
-        ValueDeletable::Value(data.slice(..value_len))
-    };
-
-    let mut key = BytesMut::with_capacity(key_prefix_len + key_suffix_len);
-    key.extend_from_slice(&first_key[..key_prefix_len]);
-    key.extend_from_slice(&key_suffix);
-
-    Ok(KeyValueDeletable {
-        key: key.into(),
-        value,
-        attributes: RowAttributes {
-            ts: meta.timestamp,
-            expire_ts: meta.expire_ts,
-        },
-    })
-}
+pub(crate) struct SstRowCodecV1 {}
 
-struct RowMetadata {
-    flags: RowFlags,
-    timestamp: Option<i64>,
-    expire_ts: Option<i64>,
-}
+impl SstRowCodecV1 {
+    pub fn new() -> Self {
+        Self {}
+    }
 
-fn decode_meta(row_features: &[RowFeature], data: &mut Bytes) -> Result<RowMetadata, SlateDBError> {
-    let mut meta = RowMetadata {
-        flags: RowFlags::empty(),
-        timestamp: None,
-        expire_ts: None,
-    };
-    for attr in row_features {
-        match attr {
-            RowFeature::Flags => match decode_row_flags(data.get_u8()) {
-                Ok(flags) => {
-                    meta.flags = flags;
-                }
-                Err(e) => return Err(e),
-            },
-            RowFeature::Timestamp => meta.timestamp = Some(data.get_i64()),
-            RowFeature::ExpireAtTs => {
-                let expire_ts = data.get_i64();
-                meta.expire_ts = if expire_ts == NO_EXPIRE_TS {
-                    None
-                } else {
-                    Some(expire_ts)
-                };
+    pub fn encode(&self, output: &mut Vec<u8>, row: &SstRowEntry) {
+        match &row.key {
+            SstRowKey::SuffixOnly { prefix_len, suffix } => {
+                output.put_u16(*prefix_len as u16);
+                output.put_u16(suffix.len() as u16);
+                output.put(suffix.as_ref());
+            }
+            _ => unreachable!("only suffix only key is supported on encode"),
+        }
+
+        let flags = {
+            let mut flags = RowFlags::empty();
+            if row.expire_ts.is_some() {
+                flags |= RowFlags::HAS_EXPIRE_TS;
+            }
+            if row.create_ts.is_some() {
+                flags |= RowFlags::HAS_CREATE_TS;
+            }
+            if row.value.as_option().is_none() {
+                flags |= RowFlags::Tombstone;
             }
+            flags
+        };
+
+        // encode seq & flags
+        output.put_u64(row.seq);
+        output.put_u8(flags.bits());
+        if flags.contains(RowFlags::Tombstone) {
+            return;
         }
+
+        // encode expire & create ts
+        if flags.contains(RowFlags::HAS_EXPIRE_TS) {
+            output.put_i64(row.expire_ts.unwrap_or(0));
```

Comment:
we should probably panic or return an error result in the situation where the row flags indicate we have expire_ts and yet there is none. we don't want to accidentally set the expire ts to `0` if it was passed in as `None`, that seems dangerous (it will immediately expire, whereas I would expect a `None` value to indicate never expire)

Reply:
make sense, fixed it with an `.expect()` here, PTAL